gradoop-flink/src/main/java/org/gradoop/flink/model/impl/operators/aggregation/functions/average/AverageProperty.java

```diff
@@ -0,0 +1,142 @@
+ *   <li>The number of values added to the sum.</li>
+ * </ol>
+ * A post-processing step is necessary after the aggregation, to get the final average value.
+ * The final value will be a {@code double} value or {@link PropertyValue#NULL_VALUE null},
+ * if there were no elements aggregated (i.e. if the property was not set on any element).
+ */
+public abstract class AverageProperty extends BaseAggregateFunction
+  implements AggregateDefaultValue {
+
+  /**
+   * A property value containing the number {@code 1}, as a {@code long}.
+   */
+  protected static final PropertyValue ONE = PropertyValue.create(1L);
+
+  /**
+   * The default value used in this aggregation.
+   */
+  private final PropertyValue defaultValue = PropertyValue.create(
+    Arrays.asList(PropertyValue.create(0L), PropertyValue.create(0L)));
+
+  /**
+   * The key used to read the value to aggregate from.
+   */
+  private final String propertyKey;
+
+  /**
+   * Creates a new instance of a base average aggregate function with a default aggregate property
+   * key (will be the original property key with prefix {@code avg_}).
+   *
+   * @param propertyKey The key of the property to aggregate.
+   */
+  public AverageProperty(String propertyKey) {
+    this(propertyKey, "avg_" + propertyKey);
+  }
+
+  /**
+   * Creates a new instance of a base average aggregate function.
+   *
+   * @param propertyKey          The key of the property to aggregate.
+   * @param aggregatePropertyKey The propertyKey used to store the aggregate.
+   */
+  public AverageProperty(String propertyKey, String aggregatePropertyKey) {
+    super(aggregatePropertyKey);
+    this.propertyKey = Objects.requireNonNull(propertyKey);
+  }
+
+  @Override
+  public PropertyValue aggregate(PropertyValue aggregate, PropertyValue increment) {
+    List<PropertyValue> aggregateValue = aggregate.getList();
+    List<PropertyValue> incrementValue = increment.getList();
+    PropertyValue sum = PropertyValueUtils.Numeric.add(aggregateValue.get(0),
+      incrementValue.get(0));
+    PropertyValue count = PropertyValueUtils.Numeric.add(aggregateValue.get(1),
+      incrementValue.get(1));
+    aggregateValue.set(0, sum);
+    aggregateValue.set(1, count);
+    aggregate.setList(aggregateValue);
+    return aggregate;
+  }
+
+  @Override
+  public PropertyValue getDefaultValue() {
+    return defaultValue;
+  }
+
+  @Override
+  public PropertyValue getIncrement(EPGMElement element) {
+    PropertyValue value = element.getPropertyValue(propertyKey);
+    return value == null ? defaultValue : asInternalAggregate(value);
+  }
+
+  /**
+   * Calculate the average from the internally used aggregate value.
+   *
+   * @param result The result of the aggregation step.
+   * @return The average value (or null, if the there were no element to get the average of).
+   */
+  @Override
+  public PropertyValue postAggregate(PropertyValue result) {
+    List<PropertyValue> value = result.getList();
```

Comment:
I don't think thats necessary, since the list is created earlier in `getIncrement`. Every step in between is clearly defined in `Aggregation` and there is nothing in between accessing the value. Even if some PropertyValue with the same key were to exist before, it would be overwritten with the list.

Reply:
But there is no fix connection to `getIncrement`. Since this is a public function where `PropertyValue result` can be an Integer, a String or NULL, you have to check that `result` is exactly whar you are expect. In my opinion, assumptions are always bad, especially in the case of public functions that can be accessed by anyone.

Reply:
Since this isn't meant for public access anyway, it would in my opinion be better to somehow restrict access to the operators using it.

Just because our package structure needs this to be public, we imho don't have to check twice to get better error messages than the default ones. For example PropertyValues `get` would return an exception like "tried to access this String as List" anyway.

Reply:
You are persistent ^^ 
_don't have to check twice_ -> OK, `get` will check it, so the first condition is only for a better error message. For me it is a common pattern to check e.g. before accessing a list entry at `1` that the list has this entry to avoid a `IndexOutOfBoundsException`. If you have a very flexible parameter type like the PropertyValue here, we should check that we get what we expect.

Sure, we do not have to be too extreme about these things, but we should be careful to make our code more robust. Part of that is to include such conditions. We can discuss about that in our meetup.